core/templates/domain/statistics/LearnerActionObjectFactory.ts

```diff
@@ -67,21 +74,17 @@ export class LearnerAction {
   }
 }
 
+export type GenericLearnerAction = (
+  LearnerAction<'ExplorationStart' | 'AnswerSubmit' | 'ExplorationQuit'>);
+
 @Injectable({
   providedIn: 'root'
 })
 export class LearnerActionObjectFactory {
-  /**
-   * @property {string} actionType - type of an action
-   * @property {Object.<string, *>} actionCustomizationArgs - customization
-   *   dict for an action
-   * @property {number} [schemaVersion=LEARNER_ACTION_SCHEMA_LATEST_VERSION]
-   *   - schema version of the class instance.
-   * @returns {LearnerAction}
-   */
-  createNew(
-      actionType: string, actionCustomizationArgs: any,
-      schemaVersion: number): LearnerAction {
+  createNew<ActionType>(
```

Comment:
Let's not avoid exposing the generic type parameter. It should remain an implementation detail. Instead, we should use a switch-statement here so that we can catch new `ActionType` values as a compiler error.

Reply:
We can remove the type parameter from createFromBackendDict part using switch case. But I was unable to do that here in `createNew`. Can you think of any method?

Also, you are saying not to expose type parameters because this users have to know the types of `action type`. But creating a different class for each one instead also does the almost same thing. Users will have to know the different classes. 

> Instead, we should rely on the switch-statement to return the correct type. That will allow us to catch new ActionType values as an error.
That work is done by typescript, isn't it?

Also, I'm fine with both the approaches but the main problem would be functions like `createNew`. If we can find a way to define types for createNew function and that too without type parameters and ts-ignore. I think we can go with this approach. Please help me here. I'm not able to think of any nice approach for the `createNew` function.

Reply:
Let's just make multiple createNew functions then, e.g.: `createNewAnswerSubmitAction`, `createNewExplorationQuitAction`, `createNewExplorationStartAction`.

> Also, you are saying not to expose type parameters because this users have to know the types of action type. But creating a different class for each one instead also does the almost same thing. Users will have to know the different classes.

Yes but concrete types are stronger than a generic type. You can pass a fake string as a generic type to create an invalid value, you can not do the same with a concrete type.

> Instead, we should rely on the switch-statement to return the correct type. That will allow us to catch new ActionType values as an error.

No, typescript won't produce an error unless the discriminated union is assigned to a `never` object. When that happens, unsupported types will create an error: blog post about it (https://medium.com/@KevinBGreene/surviving-the-typescript-ecosystem-type-guards-and-discriminated-unions-3b833c7aff04#91c4): 
![image](https://user-images.githubusercontent.com/5094060/85325946-72ab1780-b49a-11ea-9099-bc0f1d5f3807.png)

Reply:
Okay, thanks! done